### PR TITLE
Added dockerfile for 3.8.35, 3.9.22, 3.10.7 + build system using actions

### PR DESCRIPTION
**Pull request by nithinmohantk**
RabbitMQ Dockerfile maintenance releases for micdenny/rabbitmq, to keep it up-to-date according to :https://www.rabbitmq.com/changelog.html

Please build and publish the Docker images or if you would like to grant me co-ownership or maintainer status to the repo, I would be able to build and publish the docker image. I could also work on an GitHub actions pipeline for automation in a later time.  

- Added Dockerfile for 3.8.35 
- Added Dockerfile for 3.9.22
- Added Dockerfile for 3.10.7
- Updated Dockerfile for latest to use ltsc2019, instead of 2022 to support lower environments onwards
Please review merge the changes at your convenience

### CODE REVIEW

File: rabbitmq/windows/3.10.5/Dockerfile
Comment (micdenny) [inline rabbitmq/windows/3.10.5/Dockerfile:1]: please change in ltsc2022

File: rabbitmq/windows/3.9.20/Dockerfile
Comment (micdenny) [inline rabbitmq/windows/3.9.20/Dockerfile:1]: please change in ltsc2022

File: rabbitmq/windows/latest/Dockerfile
Comment (micdenny) [inline rabbitmq/windows/latest/Dockerfile:1]: please leave ltsc2022

**Review by micdenny — COMMENTED** [review]
in the meanwhile also 3.8.35 and 3.10.6 has been released, would you mind to add those as well?

File: rabbitmq/windows/3.9.20/Dockerfile
Comment (nithinmohantk) [reply rabbitmq/windows/3.9.20/Dockerfile:1]: Unfortunately that will lockdown the image to only host on ltsc2022 as minimum.
ltsc2019 will ensure it can run on both.

Still lots of companies are using 2019 hosts including AKS.

Shouldn't we keep it as is?

**Comment by nithinmohantk** [comment]
> in the meanwhile also 3.8.35 and 3.10.6 has been released, would you mind to add those as well?

Updated all of them to latest 3.8.15. 3.10.7, 3.9.22

**Comment by micdenny** [comment]
@nithinmohantk so you changed your mind regarding the windows server version? now I see all stick to ltsc2022 right?

**Comment by nithinmohantk** [comment]
> @nithinmohantk so you changed your mind regarding the windows server version? now I see all stick to ltsc2022 right?

It's just you were preferring ltsc2022, so went ahead and made changes as requested, as PR has been sitting stale for weeks.  :-) Personally, I will start from ltsc2019 onwards. Anyways thanks.

**Review by micdenny — CHANGES_REQUESTED** [review]
we should also remember to change the readme, and add the tags we are adding here

**Comment by nithinmohantk** [comment]
> we should also remember to change the readme, and add the tags we are adding here

Done. Plz review.

**Comment by micdenny** [comment]
> It's just you were preferring ltsc2022, so went ahead and made changes as requested, as PR has been sitting stale for weeks. :-) Personally, I will start from ltsc2019 onwards. Anyways thanks.

my last comment was:

"The previous images was based on 2022, for me it's ok to leave 2019 as far as they will support who was using the previous image 2022"

so basically was ok to downgrade to 2019, as far as it will still works for the one was using the old image

but probably the best choice is to create different tag for different OS version isn't it?

about the tag without os name, I would leave the latest, then I will add tag for os version as well, to end up with something like this that should work for everyone out there:

- 3.8.35
- 3.8.35-servercore-ltsc2019
- 3.8.35-servercore-ltsc2022
- 3.9.22
- 3.9.22-servercore-ltsc2019
- 3.9.22-servercore-ltsc2022
- 3.10.7
- 3.10.7-servercore-ltsc2019
- 3.10.7-servercore-ltsc2022

what do you think about?

**Comment by nithinmohantk** [comment]
> > It's just you were preferring ltsc2022, so went ahead and made changes as requested, as PR has been sitting stale for weeks. :-) Personally, I will start from ltsc2019 onwards. Anyways thanks.
> 
> my last comment was:
> 
> "The previous images was based on 2022, for me it's ok to leave 2019 as far as they will support who was using the previous image 2022"
> 
> so basically was ok to downgrade to 2019, as far as it will still works for the one was using the old image
> 
> but probably the best choice is to create different tag for different OS version isn't it?
> 
> about the tag without os name, I would leave the latest, then I will add tag for os version as well, to end up with something like this that should work for everyone out there:
> 
> * 3.8.35
> * 3.8.35-servercore-ltsc2019
> * 3.8.35-servercore-ltsc2022
> * 3.9.22
> * 3.9.22-servercore-ltsc2019
> * 3.9.22-servercore-ltsc2022
> * 3.10.7
> * 3.10.7-servercore-ltsc2019
> * 3.10.7-servercore-ltsc2022
> 
> what do you think about?

That's perfectly doable we can use ARGS for Dockerfile to maintain a single docker file per version like below. 
`# ltsc2019 or ltsc2022
ARG SERVER_VERSION = 'ltsc2022'  
FROM mcr.microsoft.com/windows/servercore:${SERVER_VERSION}`

Will shortly issue a PR updating README.md with build commands etc.

**Comment by nithinmohantk** [comment]
@micdenny Pushed in a new approach. Please review README.md for docker commands to build respectively

File: README.md
Comment (micdenny) [inline README.md:25]: at this point I would write only a couple of sample, instead of trying adding all the possible combination, so something like this, but maybe written better as you think could be better and concise:

## RabbitMQ
```powershell
docker build -t micdenny/rabbitmq-windows:x.y.z ./rabbitmq/windows/x.y.z
docker build -t micdenny/rabbitmq-windows:x.y.z ./rabbitmq/windows/latest
```

### Build for specific Windows Server versions
```powershell
docker build -t micdenny/rabbitmq-windows:x.y.z-servercore-ltsc2019 ./rabbitmq/windows/x.y.z --build-arg SERVER_VERSION=ltsc2019 
docker build -t micdenny/rabbitmq-windows:x.y.z-servercore-ltsc2012 ./rabbitmq/windows/x.y.z --build-arg SERVER_VERSION=ltsc2022 
```

## Redis
```powershell
docker build -t micdenny/redis-windows:x.y ./redis/windows/x.y
```

**Review by micdenny — CHANGES_REQUESTED** [review]
the problem with the ARGS could be with the automated build set on docker hub, but connecting to my account, it seems they removed the automated build as they want me to pay for it, and I would need to do this: https://www.docker.com/community/open-source/application/

so at this stage I would need to build locally the images and push it manually 😱 

I try to apply to the open source community, but then I think the ARG could not be used like that, I think we have to split in multiple files

File: .github/ci-windows-latest.yaml
Comment (micdenny) [inline .github/ci-windows-latest.yaml:34]: is there something like [matrix strategy of azure devops pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#multi-job-configuration)? do avoid having to copy and paste for every version? or something like [each keyword](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#each-keyword)?

File: .github/ci-windows-latest.yaml
Comment (micdenny) [inline .github/ci-windows-latest.yaml:24]: is `secrets` the name that should be used here?

image

**Review by micdenny — CHANGES_REQUESTED** [review]
using github actions is fine to me, even because I don't know if and when they will accept my request on docker hub and also with github actions we have more fine control

File: .github/workflows/ci-windows-latest.yaml
Comment (micdenny) [inline .github/workflows/ci-windows-latest.yaml:1]: Should it be better for the CI to take in consideration all the versions? so also 3.6.12, 3.6.9, 3.8.16, 3.9.8?

**Review by micdenny — CHANGES_REQUESTED** [review]

File: .github/ci-windows-latest.yaml
Comment (micdenny) [reply .github/ci-windows-latest.yaml]: we can also do it later

File: .github/workflows/ci-windows-latest.yaml
Comment (micdenny) [inline .github/workflows/ci-windows-latest.yaml:58]: Do we need this if above there's already a docker login?

**Comment by nithinmohantk** [comment]
Updated the Github pipeline to build all rabbitmq dockerfiles including older ones. 
Kindly please approve it or discard it. It's been sitting on the approval queue for quite a long, by the time there will be another new release from RabbitMQ. 
Your repo is most widely used in dockerhub for windows. It would be a great help for man kind.

**Comment by micdenny** [comment]
> Updated the Github pipeline to build all rabbitmq dockerfiles including older ones.
> Kindly please approve it or discard it. It's been sitting on the approval queue for quite a long, by the time there will be another new release from RabbitMQ.

I'll try to close it this this week 🙏 

> It would be a great help for man kind.

😬

**Review by micdenny — APPROVED** [review]
I'm going to set up the docker hub credential and give it a try

**Comment by nithinmohantk** [comment]
Cool. Need to setup like this, since the Github Action will only trigger from main or master, you might need to merge the code, as you only have write access. 

image

**Comment by micdenny** [comment]
I'm checking from your fork, and I see is getting error on 
image

is this pushing on your docker hub account right now? so I can check first there before merging here

**Comment by nithinmohantk** [comment]
> probably this, isn't it? 😄 https://hub.docker.com/r/thingxcloud/rabbitmq-windows

Yes. It is used for testing builds. I am going to remove that repo soon.

**Comment by micdenny** [comment]
I wait to see if your last merge on your master is greening, then I close this PR

image

**Comment by nithinmohantk** [comment]
> I'm checking from your fork, and I see is getting error on image
> 
> is this pushing on your docker hub account right now? so I can check first there before merging here

Fork will fail because the token on it is valid for my test dockerhub repo only. You will have to setup new tokens in your main repo.  https://github.com/micdenny/docker-samples

**Comment by nithinmohantk** [comment]
> I wait to see if your last merge on your master is greening, then I close this PR
> 
> image

This will always fail from my repo, because it is trying to push to [micdenny/rabbitmq-windows]. You need to merge the changes to your repo, and implement secrets accordingly to that Github Actions kicks off from your repo, will push new images to [micdenny/rabbitmq-windows]

**Comment by micdenny** [comment]
> Hope it is making sense to you.

yes that's clear, I was hoping to see a build and push to your repo all green, but I'm going to close this in master and if there's a problem we will hotfix it

**Comment by nithinmohantk** [comment]
> > Hope it is making sense to you.
> 
> yes that's clear, I was hoping to see a build and push to your repo all green, but I'm going to close this in master and if there's a problem we will hotfix it

Manually updated my [master] top push containers to [thingxcloud] repo for testing.

**Comment by micdenny** [comment]
Ok in the meanwhile I've already configured the secrets, your build is running, I wait the green and I merge this

**Comment by micdenny** [comment]
@nithinmohantk is there a way to automatically set the overview page of the hub with the content of the README.md of this repo under https://github.com/micdenny/docker-samples/tree/master/rabbitmq? This is how the old hub build system was working, so updating the readme here also the overview was synchronized.

**Comment by micdenny** [comment]
This is what I was talking about, I already saw this error from a build of yesterday

image

Have they de-listed this version maybe?

**Comment by nithinmohantk** [comment]
> Have they de-listed this version maybe?
It appears yes, they might have delisted the specific version of erlang.

**Comment by micdenny** [comment]
3.6.9

image

3.6.12

image

searching if they uploaded in a different way on github

**Comment by micdenny** [comment]
ok you found it 😄 wait to see a green from your builds

**Review by micdenny — CHANGES_REQUESTED** [review]
I think you want remove this before I merge it

image

**Comment by nithinmohantk** [comment]
> I think you want remove this before I merge it
> 
> image

Will remove it after build is successful.

**Comment by nithinmohantk** [comment]
Succeeded. Will revert the yaml file now. 
image

**Review by micdenny — APPROVED** [review]

**Comment by micdenny** [comment]
good job @nithinmohantk thank you very much for helping me on this!

**Pull request merged**
